### ISSUE TITLE
SyncActorPublisher is swallowing mongo connection errors.

### ISSUE THREAD

**Issue by PiotrFras** [scullxbones/akka-persistence-mongo#156]
https://github.com/scullxbones/akka-persistence-mongo/blob/master/common/src/main/scala/akka/contrib/persistence/mongodb/MongoReadJournal.scala#L346

`SyncActorPublisher.preStart`, if the call to `initialCursor` fails with exception, the stream will enter limbo state, never erroring nor completing.

Following code demonstrates the issue:

```
import akka.actor.{ActorSystem, Props}
import akka.stream.ActorMaterializer
import akka.stream.actor.ActorPublisher
import akka.stream.scaladsl.Source

object Main {

  def main(args: Array[String]): Unit = {

    implicit val system = ActorSystem()
    implicit val mat = ActorMaterializer()

    Source.actorPublisher(Props(new SourceActor)).recover {
      case _ => "recovered from error"
    }.runForeach(println)

    println("exit!")
  }

  class SourceActor extends ActorPublisher[String] {

    override def preStart(): Unit = throw new RuntimeException("???")

    override def receive: Receive = {
      case _ =>
    }
  }

}
```

Above program never prints `recovered from error`.

This is not akka issue, `ActorPublisher` is deprecated in 2.5

**Comment by bfil** [comment]
This seems to be the case because the read journaller code does not have any error handling so it's not very resilient.

This line appears to be only executed once:

https://github.com/scullxbones/akka-persistence-mongo/blob/master/casbah/src/main/scala/akka/contrib/persistence/mongodb/CasbahPersistenceReadJournaller.scala#L137

So if the cursor fails it will never be recreated:

https://github.com/scullxbones/akka-persistence-mongo/blob/master/casbah/src/main/scala/akka/contrib/persistence/mongodb/CasbahPersistenceReadJournaller.scala#L122

So any new queries using the realtime stream will never receive new journal events.
This should be rebuilt in a way that would recover in case of a database connection drop.

@scullxbones any preference over how this could be implemented?

**Comment by scullxbones** [comment]
@PiotrFras - is this the same issue as #155?  It sounds similar.  My surprise is that I expected an exception thrown in `preStart` to cause an `ActorInitializationException`, which would propagate via streams API to cancel the stream.  Ideally that's the behavior I'd want ultimately.  I'm thinking catching the exception from `initialCursor` and use that to send a message to `self` which causes an immediate stream cancellation with the exception.  I can put that together pretty quickly, unless that was what you wanted to submit a PR on, in which case go for it.  Let me know either way.

@bfil - this looks like a related but separate issue of general robustness on the casbah read journaller.  Given those choices I'd say rewriting the `CasbahMongoJournalStream` as an actor which maintains the cursor state and retries on failure with a delay.  I'll create a second issue for this part.  I'd welcome a PR on this one.

For both issues - I'm going to want to back-port these to the akka 2.4 branch.

**Comment by bfil** [comment]
@scullxbones I guess they're all a little related, in general, error handling should be improved in the Casbah read journal, not sure about the RxMongo one, but it looks pretty similar.

In general the MongoPersistenceReadJournallingApi doesn't look very good with some methods returning `Source`s and one method accepting an `ActorRef` and returning `Unit`. This part in particular needs improvement:

```scala
override def publishEvents(): Unit = {
  implicit val ec = driver.querySideDispatcher
  Future {
    cursor().foreach { next =>
      if (next.keySet().contains(EVENTS)) {
        val events = next.as[MongoDBList](EVENTS).collect { case x: DBObject => driver.deserializeJournal(x) }
        events.foreach(driver.actorSystem.eventStream.publish)
      }
    }
  }
  ()
}
```

It wraps a blocking operation into a `Future` without recovery and then returns `Unit`, not the best approach, but I get the intent. We'll probably keep the API as it is for a PR but imo the read journalling API should expose a similar API for `allEvents` rather than a completely different one, but there might be good reasons why this is the case, if there is please let me know.

**Comment by scullxbones** [comment]
> In general the MongoPersistenceReadJournallingApi doesn't look very good with some methods returning Sources and one method accepting an ActorRef and returning Unit.

Well that's unfortunate.  Maybe I can clear that up.

> We'll probably keep the API as it is for a PR but imo the read journalling API should expose a similar API for allEvents rather than a completely different one, but there might be good reasons why this is the case, if there is please let me know.

Quite simply, this is the case to share code between the two drivers.  All of the `Current` queries are implemented by the individual drivers since the `All` queries can be treated in a driver-agnostic way via this event-bus based stream.  There's no good reason IMO to duplicate this behavior in each of the drivers.  I wanted the design to make it easier to support more drivers, rather than harder.

**Comment by scullxbones** [comment]
Fixed by #199
